PR [KOB-2052-KOB-3417] Integrating Kobiton into CircleCI and Appium (#1)
NguyenHoanTien/CircleCI-HockeyApp-Appium-ReactNative: README.md

> # Integrating Kobiton into CircleCI-HockeyApp-Appium-ReactNative development pipeline

Reviewer: ```suggestion
# Integrating Kobiton into CircleCI development pipeline
```
This guid... apply Appium Testing to ReactNative application

> # Integrating Kobiton into CircleCI - Appium Testing - ReactNative mobile application development pipeline

Reviewer: I still have wrong feeling with this title, I will keep my suggestion.
```suggestion
# Integrating Kobiton into CircleCI development pipeline
```
Then mention the Appium + ReactNative below

> Kobiton is a mobile **experience** platform that accelerates delivery and testing of mobile apps by offering manual and automated testing on real devices, in the cloud or on-premise.

Reviewer: remove this part

> In application development, especially mobile application development, perform testing on **real** devices before releasing is a crucial process, but also a costly process. In fact, only a few number of developers can be able to perform testing on real devices, but only a very limited range of devices.

Reviewer: Make it short and brief, directly to the point

Reviewer: ???

> By integrating Kobiton into mobile application development pipeline, developers can be able to execute manual and automation tests on a wide range of real devices, thus can greatly improve the quality of their applications.

Reviewer: 1. apply automation testing
2. Using source code control for saving test script.

> By integrating Kobiton into your mobile application development pipeline, you can perform automation testing on a wide range of devices, which can greatly improve the quality of your applications. Also, if you already have your test scripts written in Appium, you can easily migrate and execute them on Kobiton. The integration is demonstrated in the diagram below:

Reviewer: - Test on ReactNative application
- Appium testing script
- Using version control source (github) 
-> Explaination

Below is the diagram demonstrating how Kobiton integrates with your mobile application development pipeline.

<img src="docs/assets/diagram.svg"/>

> This guide will give you an in-depth demonstration (with examples) of how to integrate Kobiton with CircleCI. Refer to [our documentation](./docs/README.md) for more details.

Reviewer: remove this line

> From above diagram, by integrating Kobiton into your mobile development pipeline using CircleCI as a CI tool, CircleCI will automatically pull the automation test scripts from a GitHub repository and execute automation tests on Kobiton on triggered or every time a commit is pushed.

Reviewer: Merge with above part

# Integrating Kobiton into CircleCI-HockeyApp-Appium-ReactNative development pipeline

One of the biggest problem in mobile application development is the limitation of number of real devices that developers can be able to test their application on before release, which leads to the fact that some users with untested devices can/may experience unwanted errors.

> 

Reviewer 2: the term "Users with untested devices" very confused!

Reviewer 2: "Unwanted errors" are very informal

# Integrating Kobiton into CircleCI mobile application development pipeline

In application development, especially mobile application development, perform testing on **real** devices before releasing is a crucial, but also costly process, which means that only few developers are able to perform testing on real devices, but only a very limited range of devices.

> By integrating Kobiton into your mobile application development pipeline, you can perform automation testing on a wide range of devices, which can greatly improve the quality of your applications.

Reviewer: ```suggestion
By integrating Kobiton into your mobile application development pipeline, CircleCI will able to execute an automation testing on real devices, which help improve development process and product quality. 
```

This guide is applicable if you are using these tools and technologies in your pipeline:
- ReactNative as a technology to develop the application with.
- Appium as a framework to develop the test scripts using.
- CircleCI as a CI/CD tool.
- GitHub as a VCS provider to store your test scripts on.

> As shown in the diagram below, by integrating Kobiton into your mobile development pipeline, CircleCI will automatically pull the automation test scripts from GitHub repository and execute them on Kobiton on triggered or every time a commit is pushed.

Reviewer: "by integrating Kobiton into your mobile development pipeline"
Duplicate

> As shown in the diagram below, by integrating Kobiton into your mobile development pipeline, CircleCI will automatically pull the automation test scripts from GitHub repository and execute them on Kobiton on triggered or every time a commit is pushed.

Reviewer: Renove this part, I don't think this is necessary

This document will guide you how to use CircleCI to execute automation test on Kobiton devices everytime you push a commit to GitHub.

## Table of contents
- [A. Preparation](#a-preparation)
    - [1. Prerequisites](#1-prerequisites)
    - [2. Getting Kobiton Username and API key](#2-getting-kobiton-username-and-api-key)
    - [3. Samples](#3-samples)
- [B. Setup](#b-setup)
    - [Purposes](#purposes)
    - [1. Setting Kobiton Username and API Key](#1-setting-kobiton-username-and-api-key)
    - [2. Getting Kobiton device desired capabilities](#2-getting-kobiton-device-desired-capabilities)
    - [3. Setting desired capabilities](#3-setting-desired-capabilities)
    - [4. CircleCI configuration file](#4-circleci-configuration-file)
    - [5. Configuring HockeyApp WebHook listener](#5-configuring-hockeyapp-webhook-listener)
- [C. Execution](#c-execution)
    - [1. Executing the automation test](#1-executing-the-automation-test)
    - [2. Fetching test session data using Kobiton REST API](#2-fetching-test-session-data-using-kobiton-rest-api)
- [D. Feedback](#d-feedback)

## A. Preparation

### 1. Prerequisites
- A server for receiving WebHooks.

>     - Can be virtual, bare-metal or container, etc.

Reviewer 3: I don't think this line is necessary unless there are exception(s) in configuring the server.

In this guide, we will be using the provided examples. Follow steps below to get started:

1. Fork this repository https://github.com/kobiton/CircleCI-HockeyApp-Appium-ReactNative

2. Synchronize the forked repository with CircleCI. If you don't know how to do it, refer to instruction at [CircleCI document](https://circleci.com/docs/2.0/getting-started/)

> Note : At first synchronization, the build process will be failed to execute. This is normal, just ignore the `Build Failed` message.

3. Clone the forked repository as we will be modifying it later.

## B. Setup
### 1. Setting Kobiton Username and API key

In your project configuration in CircleCI, add these two environment variables :
- KOBITON_USERNAME - Your Kobiton's username.
- KOBITON_API_KEY - Your Kobiton's API Key.

Your `Environment Variables` window should look like this

![Environment variables](./docs/assets/env-variables-after-add.png)

### 2. Getting Kobiton device desired capabilities
In order to interact with a specific device in Kobiton, its corresponding desired capabilities needs to be supplied. 

> In the provided sample automation test script, we have pre-configured them to execute automation test of the provided application(s) from `APP_URL` environment variable, and execute the tests on one of available devices which name begins with `iPhone` or `Android` if **ONLY** platform name is supplied in desired capabilities.

Reviewer 3: What do you mean by `name`? Is it the same as `deviceName` ? If so, having `deviceName`  begins with `Android` doesn't trigger an auto test on android device(s).

**How to get desired capabilities**

In this tutorial, we will be showing how to get desired capabilities for executing automation test on `Pixel 2 XL` running `Android 8.1.0`.

1. Go to [https://portal.kobiton.com/login](https://portal.kobiton.com/login) and login with your Kobiton account.
2. Click **"Devices"** at the top of the window.

![Devices button](./docs/assets/devices-button.png)

3. Hover over the device you want to run your test script on, click the gear button.

![Automation Settings button](./docs/assets/gear-button.png)

4. In the `Automation Settings` popup:
- In `Language` section, choose `NodeJS`.

- In `App Type` section, choose `Hybrid/Native from Apps`.

![Example Android DesiredCaps](./docs/assets/desired-caps-example-android.png)

### 3. Setting desired capabilities

In `samples/automation-test` folder, we have provided two sample scripts for executing automation test on Kobiton iOS and Android devices.

As mentioned above, the desired capabilities in provided automation test scripts have already been pre-configured.

> If you wish to use default configuration, add only a variable named `KOBITON_DEVICE_PLATFORM_NAME` with value `iOS` if you want to execute on iOS or `Android` if on Android.

Reviewer 3: We don't need this line.

If you have chosen custom desired capabilities, please follow steps below to apply the collected desired capabilities.

**Setting desired capabilities**

Add there environment variables to CircleCI Environment Variables configuration window:
- `KOBITON_DEVICE_NAME`: Kobiton device name (`deviceName` in the generated desired capabilities).

> - `KOBITON_DEVICE_PLATFORM_NAME`: Kobiton device paltform name (`platformName` in the generated desired capabilities).

Reviewer 3: `platform`

- `KOBITON_DEVICE_PLATFORM_VERSION`: Kobiton device platform version (`platformVersion` in the generated desired capabilities).

For example, if desired capabilities for executing automation test on `Pixel 2 XL` running `Android 8.1.0`:

```javascript
const desiredCaps = {
  sessionName:        'Automation test session',
  sessionDescription: 'Demo Automation Test on Android', 
  deviceOrientation:  'portrait',  
  captureScreenshots: true, 
  app:                '<APP_URL>',
  deviceGroup:        'KOBITON', 
  deviceName:         'Pixel 2 XL',
  platformVersion:    '8.1.0',
  platformName:       'Android' 
}
```

Then the corresponding environment variables set in CircleCI should be

![Example Android DesiredCaps](./docs/assets/desired-caps-circleci-env.png)

> More information about Desired Capabilities and its parameters can be found in https://docs.kobiton.com/automation-testing/desired-capabilities-usage/

> #### 4. CircleCI configuration file

Reviewer 3: `### 4.`

In the provided CircleCI configuration file `/.circleci/config.yml`, the execution environment has already been set to using the mentioned container.

```yaml
docker:
    - image: circleci/node:latest
```

> For more information about how to execute automation test(s) on Kobiton, you can visit:
> - Kobiton documentation: https://docs.kobiton.com/automation-testing/automation-testing-with-kobiton/
> - Kobiton's samples for automation test scripts in other languages: https://github.com/kobiton/samples

**Automation test script execution**

To execute the automation test script, just add a `run` section with command `npm run <SCRIPT_FILENAME>`. By default, it has already been added to the provided CircleCI configuration file.

Example of executing the provided automation script:

```yaml
- run:
    name: app-auto-testing
    command: npm run automation-test-script
    working_directory: samples/automation-test
```

### 5. Configuring HockeyApp WebHook listener

> Kobiton has provided script for setting up a small, basic server for receiving WebHook notifications from HockeyApp and triggering CircleCI to execute automation test of that application on Kobiton. The script will load settings from a configuration file and start a WebHook listener. This script must be executed on the server mentioned in **Prerequisites**.

Reviewer 3: Make a hyperlink for `**Prerequisites**`

> Kobiton has provided script for setting up a small, basic server for receiving WebHook notifications from HockeyApp and triggering CircleCI to execute automation test of that application on Kobiton. The script will load settings from a configuration file and start a WebHook listener. This script must be executed on the server mentioned in **Prerequisites**.

Reviewer 3: Please highlight this (bolding, write in a new line,...). This line is essential yet is too easy to miss.
```
This script must be executed on the server mentioned in **Prerequisites**.
```

- `<YOUR_PROJET_ORGANIZATION>`: Your repository organization.
- `<YOUR_PROJECT_REPOSITORY>`: Your repository name.
- `<PROJECT_BRANCH>`: Branch of the repository that contain the automation test script you want to execute (default: master).
3. Execute the provided script using the below command:
```bash
ruby webhook_server.rb
```

4. In your application configuration in HockeyApp, change WebHook server address to your corresponding server address, port.

5. Send a ping request from HockeyApp. If your server console output contains `Ping request received`, then you have successfully configured your server to listen for WebHook from HockeyApp.

## C. Execution
### 1. Executing the automation test

Simply push your modified test script(s) and CircleCI configuration file `/.circleci/config.yml` to your GitHub repository. CircleCI will automatically configure the execution environment, install missing dependencies and execute the test script in `/samples` based on the provided configuration file `/.circleci/config.yml`.

After you have pushed, the build process on CircleCI will end with error. That is normal, just simply ignore it.

> To execute the provided automation test script(s), simply upload your application package to HockeyApp. HockeyApp will process the uploaded file, send notification to the WebHook server, the server will then execute the automation test script on CircleCI with the URL received from HockeyApp WebHook notification.

Reviewer 3: Please note that users can upload new app or new version of the app, please explain it more clearly.

> To execute the provided automation test script(s), simply upload your application package to HockeyApp. HockeyApp will process the uploaded file, send notification to the WebHook server, the server will then execute the automation test script on CircleCI with the URL received from HockeyApp WebHook notification.

Reviewer 3: I think we need a picture of the server getting webhook request.

Your test execution process can be viewed on CircleCI

![CircleCI build execution title](./docs/assets/test-title.png)

![CircleCI build execution report](./docs/assets/test-details.png)

Your test session progress can be viewed on Kobiton website. Follow these steps below

1. Go to [https://portal.kobiton.com/sessions](https://portal.kobiton.com/sessions), login with your Kobiton account.

2. You will see your executed sessions and their statuses.

![Kobiton session execution](./docs/assets/session-running-kobiton.png)

> 3. Click on any session to view its details, commands.

Reviewer 3: `Click on any session for detailed information.`

Reviewer 3: Also this line is kinda easy to miss (since it's placed below a picture).

> # CircleCI-HockeyApp-Appium-ReactNative

Reviewer 4: you need Readme.MD file to introduce this guideline

    - [2. Fetching test session data using Kobiton REST API](#2-fetching-test-session-data-using-kobiton-rest-api)
- [D. Feedback](#d-feedback)

## A. Preparation

### 1. Prerequisites
- A server for receiving WebHooks.
    - Can be virtual, bare-metal or container, etc.
    - **MUST** be visible to the internet.
    - Has Ruby 2.x installed.
    > If you haven't installed Ruby on your server, please refer to [Ruby Installation documentation](https://www.ruby-lang.org/en/documentation/installation/) on how to install Ruby on corresponding OSes.
    
### 2. Getting Kobiton Username and API key
Kobiton Username and API key are required for authenticating with Kobiton.

> If you don't have a Kobiton account, visit https://portal.kobiton.com/register to create one.

To get your Kobiton Username and API Key, follow instruction at `IV. Configure Test Script for Kobiton` section on [our blog article](https://kobiton.com/blog/tutorial/parallel-testing-selenium-webdriver/).

### 3. Samples
In this repository, we have already provided samples for executing automation test in Kobiton:
- Script for executing automation test on Kobiton iOS devices (`/samples/automation-test/ios-app-test.js`).

> - Script for executing automation test on Kobiton Android devices (`/samples/automation-test/android-app-test.js`).

Reviewer 4: Remove 2 lines
-> Script for executing automation test on Kobito devices (`/samples/automation-test/automation-test-scripts.js`).

In this guide, we will be using the provided examples. Follow steps below to get started:

1. Fork this repository https://github.com/kobiton/CircleCI-HockeyApp-Appium-ReactNative

2. Synchronize the forked repository with CircleCI. If you don't know how to do it, refer to instruction at [CircleCI document](https://circleci.com/docs/2.0/getting-started/)

> Note : At first synchronization, the build process will be failed to execute. This is normal, just ignore the `Build Failed` message.

3. Clone the forked repository as we will be modifying it later.

## B. Setup
### 1. Setting Kobiton Username and API key

In your project configuration in CircleCI, add these two environment variables :
- KOBITON_USERNAME - Your Kobiton's username.
- KOBITON_API_KEY - Your Kobiton's API Key.

Your `Environment Variables` window should look like this

![Environment variables](./docs/assets/env-variables-after-add.png)

### 2. Getting Kobiton device desired capabilities
In order to interact with a specific device in Kobiton, its corresponding desired capabilities needs to be supplied. 

> In the provided sample automation test script, we have pre-configured them to execute automation test of the provided application(s) from value of `APP_URL` environment variable, and check if **ONLY** the `platformName` is supplied; if it is supplied with `iOS` or `Android`, the test will be executed on one of the available devices names of which begin with `iPhone` or `Android`. If you want to use other specific device(s), follow the instruction below to get the corresponding desired capabilities for that device(s).

Reviewer 4: I think this line is confusing

The provided sample automation test script is pre-configured to execute automation test of application(s) requested from WebHook server. It will execute on device(s) matching the provided desired capabilities or on one of any available devices whose names begin with `iPhone` if being executed on `iOS` or `Galaxy` if on `Android`.

> 

Reviewer 4: if users want to execute test on a device, users need to note values of deviceName, platformversion and platformName to prepare next steps.
